File: app/src/main/java/com/yfbx/widgets/java/selector/CheckColorSelect.java
package com.yfbx.widgets.java.selector;

import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.support.annotation.Nullable;
import android.text.TextPaint;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Checkable;

import com.yfbx.widgets.R;

/**
 * Author:Edward
 * Date:2018/10/26
 * Description:
 */

public class CheckColorSelect extends View implements Checkable {

    private TextPaint paint;
    private float width;
    private float height;

    private ColorStateList colorStateList;
    private boolean checked;
    private String text;

    private static final int[] CHECKED_STATE_SET = {android.R.attr.state_checked};


    public CheckColorSelect(Context context) {
        this(context, null);
    }

    public CheckColorSelect(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public CheckColorSelect(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        this(context, attrs, defStyleAttr, 0);
    }

    public CheckColorSelect(Context context, @Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        getAttr(context, attrs);
        init();
    }

    private void getAttr(Context context, AttributeSet attrs) {
        TypedArray array = context.obtainStyledAttributes(attrs, R.styleable.CheckColorSelect);
        checked = array.getBoolean(R.styleable.CheckColorSelect_android_checked, false);
        text = array.getString(R.styleable.CheckColorSelect_android_text);
        colorStateList = array.getColorStateList(R.styleable.CheckColorSelect_android_textColor);
        array.recycle();
    }


    private void init() {
        paint = new TextPaint();
        paint.setTextSize(sp2px(14));
        paint.setStyle(Paint.Style.FILL);
        paint.setAntiAlias(true);
    }


    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        Rect textBounds = getTextBounds();
        //width
        if (MeasureSpec.getMode(widthMeasureSpec) == MeasureSpec.EXACTLY) {
            width = MeasureSpec.getSize(widthMeasureSpec);
        } else {
            width = textBounds.width() + getPaddingLeft() + getPaddingRight();
        }

        //height
        if (MeasureSpec.getMode(heightMeasureSpec) == MeasureSpec.EXACTLY) {
            height = MeasureSpec.getSize(heightMeasureSpec);
        } else {
            height = textBounds.height() + getPaddingTop() + getPaddingBottom();
        }
        setMeasuredDimension((int) width, (int) height);//存在精度损失
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        height = h;
        width = w;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        canvas.translate(0, height / 2);
        int color = colorStateList.getColorForState(getDrawableState(), colorStateList.getDefaultColor());
        paint.setColor(color);
        Rect textBounds = getTextBounds();
        canvas.drawText(text, getPaddingLeft(), -textBounds.centerY(), paint);

    }

    @Override
    protected int[] onCreateDrawableState(int extraSpace) {
        int[] states = super.onCreateDrawableState(extraSpace + 1);
        if (isChecked()) {
            mergeDrawableStates(states, CHECKED_STATE_SET);
        }
        return states;
    }

    @Override
    protected void drawableStateChanged() {
        super.drawableStateChanged();
        invalidate();
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        return event.getAction() != MotionEvent.ACTION_UP || performClick();
    }

    @Override
    public boolean performClick() {
        toggle();
        return super.performClick();
    }

    private Rect getTextBounds() {
        Rect rect = new Rect();
        paint.getTextBounds(text, 0, text.length(), rect);
        return rect;
    }

    @Override
    public void setChecked(boolean checked) {
        if (this.checked != checked) {
            this.checked = checked;
            refreshDrawableState();
        }
    }

    @Override
    public boolean isChecked() {
        return checked;
    }

    @Override
    public void toggle() {
        setChecked(!checked);
    }

    /**
     * sp 转换为 px
     */
    protected float sp2px(float value) {
        DisplayMetrics metrics = Resources.getSystem().getDisplayMetrics();
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, value, metrics);
    }

    /**
     * dp 转换为 px
     */
    private float dp2px(float value) {
        DisplayMetrics metrics = Resources.getSystem().getDisplayMetrics();
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, value, metrics);
    }
}

File: app/src/main/java/com/yfbx/widgets/widget/selector/CheckColorSelect.kt
package com.yfbx.widgets.widget.selector

import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.text.TextPaint
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.widget.Checkable
import com.yfbx.widgets.R
import com.yfbx.widgets.util.sp

/**
 * Author:Edward
 * Date:2018/10/26
 * Description:
 */

class CheckColorSelect @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0, defStyleRes: Int = 0) : View(context, attrs, defStyleAttr, defStyleRes), Checkable {


    private var paint: TextPaint = TextPaint()
    private var width = 0f
    private var height = 0f

    private var colorStateList: ColorStateList? = null
    private var checked: Boolean = false
    private var text: String? = null


    init {
        getAttr(context, attrs)
        paint.textSize = 14f.sp
        paint.style = Paint.Style.FILL
        paint.isAntiAlias = true
    }

    private fun getAttr(context: Context, attrs: AttributeSet?) {
        val array = context.obtainStyledAttributes(attrs, R.styleable.CheckColorSelect)
        checked = array.getBoolean(R.styleable.CheckColorSelect_android_checked, false)
        text = array.getString(R.styleable.CheckColorSelect_android_text)
        colorStateList = array.getColorStateList(R.styleable.CheckColorSelect_android_textColor)
        array.recycle()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val textBounds = getTextBounds()
        //width
        if (MeasureSpec.getMode(widthMeasureSpec) == MeasureSpec.EXACTLY) {
            width = MeasureSpec.getSize(widthMeasureSpec).toFloat()
        } else {
            width = (textBounds.width() + paddingLeft + paddingRight).toFloat()
        }

        //height
        if (MeasureSpec.getMode(heightMeasureSpec) == MeasureSpec.EXACTLY) {
            height = MeasureSpec.getSize(heightMeasureSpec).toFloat()
        } else {
            height = (textBounds.height() + paddingTop + paddingBottom).toFloat()
        }
        setMeasuredDimension(width.toInt(), height.toInt())//存在精度损失
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        height = h.toFloat()
        width = w.toFloat()
    }

    override fun onDraw(canvas: Canvas) {
        canvas.translate(0f, height / 2)
        val color = colorStateList!!.getColorForState(drawableState, colorStateList!!.defaultColor)
        paint.color = color
        val textBounds = getTextBounds()
        canvas.drawText(text!!, paddingLeft.toFloat(), (-textBounds.centerY()).toFloat(), paint)

    }

    override fun onCreateDrawableState(extraSpace: Int): IntArray {
        val states = super.onCreateDrawableState(extraSpace + 1)
        if (isChecked) {
            mergeDrawableStates(states, intArrayOf(android.R.attr.state_checked))
        }
        return states
    }

    override fun drawableStateChanged() {
        super.drawableStateChanged()
        invalidate()
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        return event.action != MotionEvent.ACTION_UP || performClick()
    }

    override fun performClick(): Boolean {
        toggle()
        return super.performClick()
    }

    override fun setChecked(checked: Boolean) {
        if (this.checked != checked) {
            this.checked = checked
            refreshDrawableState()
        }
    }

    override fun isChecked(): Boolean {
        return checked
    }

    override fun toggle() {
        isChecked = !checked
    }


    private fun getTextBounds(): Rect {
        val rect = Rect()
        paint.getTextBounds(text, 0, text!!.length, rect)
        return rect
    }

}
